### ISSUE TITLE
API/CLI discrepancy on hostname validation

### ISSUE THREAD

**Issue by icecrime** [docker/docker#21595]
In the CLI, running a container with a `--hostname` of more than 64 characters will fail:

```
user@ubuntu:~$ docker run -it --net=f826338a441f2f8a4685630e4fd83ad3af78c0fc --hostname=f826338a441f2f8a4685630e4fd83ad3af78c0fc_go1.f826338a441f2f8a4685630e4fd83ad3af78c0fc --name=f826338a441f2f8a4685630e4fd83ad3af78c0fc_go1 golang
docker: invalid hostname format for --hostname: f826338a441f2f8a4685630e4fd83ad3af78c0fc_go1.f826338a441f2f8a4685630e4fd83ad3af78c0fc.
See 'docker run --help'.
```

However, this seems to be allowed through the API, and fails later down the road with:

```
Error response from daemon: Cannot start container 353dc194a5cf9b34d9653b9828aa101f8f598fd2b91c4f3633ac1b5d86f4a72d: [9] System error: invalid argument
```

As far as I can tell, hostname validation happens in [`runconfig/opts/parse.go`](https://github.com/docker/docker/blob/master/runconfig/opts/parse.go#L249).

Cc @calavera @MHBauer.

Indirectly reported in #21562.

**Comment by yongtang** [comment]
Hi @icecrime, the hostname was restricted to 64 bytes in pull request #21463 (issue #21445) to conform to Linux's HOST_NAME_MAX=64. I only did the CLI parsing at that time. I can make the change to API part as well. #dibs

**Comment by thaJeztah** [comment]
Should we do it *only* in the API, to prevent duplicating the validation? @calavera? (perhaps we should document a standard approach for these things)

**Comment by yongtang** [comment]
I created a pull request #21641 to try to address this API/CLI discrepancy. Basically, the hostname validation has been moved from `runconfig/opts/parse.go` to inside `verifyContainerSettings` in `daemon/daemon.go`. 

The error message has been changed from `invalid hostname format for --hostname:` to `invalid hostname format:` as `--hostname` is passedto CLI only so it may not be suitable for API case.

Let me know if the changes meet the requirement or there are further adjustments I need to make.

cc @icecrime @calavera @MHBauer @thaJeztah.
